doc/developer/design/20231110_aws_connections.md

```diff
@@ -0,0 +1,161 @@
+
+Optionally, with `ASSUME ROLE ARN` we can also try to get temporary credentials.
+
+## Rollout and Testing
+We should put the AWS connection behind a feature flag.
+
+#### Testing During development
+Write cloudtests to test out the different AWS Connections. For the AssumeRole AWS
+Connection we should test that the expected principal and external_id is populated
+in the `mz_aws_connections` table.
+
+#### Testing after code is merged
+Switch on the feature flag for the staging environment and create an AWS Connection
+in staging followed by `VALIDATE CONNECTION`.
+
+## Future work
+
+#### Extend `VALIDATE CONNECTION` of an AWS connection to be comprehensive.
+An AWS connection can be potentially re-used across multiple services like S3 or RDS.
+Having permission to an S3 bucket might not mean that RDS access is set up correctly. So
+a validate connection will not be comprehensive and we'll probably do some additional check
+with the intended resources when we actually make use of this connection.
+
+Hence, potentially later we can extend `VALIDATE CONNECTION` sql to optionally specify an S3 or a database url,
+like, `VALIDATE CONNECTION aws_conn WITH (S3 PATH = 's3://prefix')` so the user can correctly
+validate if the policies are set up correctly on their end in AWS.
+
+Alternatively, we could have a higher level S3 connection using this AWS connection,something
+like `CREATE CONNECTION s3_conn TO S3 ( PREFIX = 'url') USING AWS CONNECTION aws_conn`.
+Validating s3_conn could verify if we are able to list the prefix, create and
+remove a file there. This seems like an overkill though and a `VALIDATE CONNECTION ... WITH`
+mentioned above would probably be better.
+
+## Open Questions
+#### What should be the `principal` for this connection?
+For an AssumeRole to work, we need to provide the principal to the user using which we'll try to assume the role.
+I had initially assumed that the principal we generate for AWS Privatelink might work, but that does not seem to be the case.
+Among many things, the principal can be
+1. Our account ID, corresponding to `arn:aws:iam::<account_id>:root`
+2. An IAM user `arn:aws:iam::<account_id>:user/<id>` we create for each user environment
+3. An IAM role `arn:aws:iam::<account_id>:role/<id>` we create for each user environment
+
+We could simplify and directly use the root account ID (Option 1). But this seems insecure that any materialize role/user
+belonging to this account would be able to assume this user's role.
+
+Option 3 will not scale, there's a maximum limit of [5000 roles](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html#reference_iam-quotas-entities) we can create per account.
+
+Snowflake uses Option 2, they create [one IAM user per user](https://docs.snowflake.com/en/user-guide/data-load-s3-config-storage-integration#step-4-retrieve-the-aws-iam-user-for-your-snowflake-account).
+If we want to do this, this will require additional work on the cloud's side.
```

Comment:
I think we'd be ok with having this be the account ID, but I worry about having the account id, with an external id and a connection id non of which are secrets. I would feel better if we could use some sort of secret/encrypted value here.

Reply:
Brought this to the cloud team sync... here are the takeaways.

First a take on the solutions we don't think we want to go with:
* using the root account - We're pretty much all against using the root account here, it'd work, but it's not as secure and not a good look (someone on the customer security / cloud team may question why they need to specify the entire account) 
* using users - this could be ok if we stored them in a secret and rotated regularly, this would give each customer a unique principal.. we don't really like the idea that these credentials could be exported and used outside of the environment cluster, we'd have to rotate *very* frequently we may also need to coordinate that role across all users of the credentials which sounds hard. 

The solution we believe is best is using a globally created role which all envd/clusterd pods can assume.

This looks better for our customers, adds some good restrictions on who could assume the customers role (only envd/clusterd), and allows for migration between environment clusters.

Reply:
@jubrad Sounds good to me 👍. Let me update the design doc with this.

Reply:
Updated!